packages/client/src/getPools.ts

```diff
@@ -51,38 +58,112 @@ export const getPools = async function (this: ZetaChainClient) {
     []
   );
 
-  const poolPromises = uniquePairs.map(async ({ tokenA, tokenB }: any) => {
-    const pair = await systemContract.uniswapv2PairFor(
+  const multicallAddress = "0xca11bde05977b3631167028862be2a173976ca11";
+  const multicallContract = new ethers.Contract(
+    multicallAddress,
+    MULTICALL3_ABI,
+    provider
+  );
+
+  const calls = uniquePairs.map(({ tokenA, tokenB }) => ({
+    callData: systemContract.interface.encodeFunctionData("uniswapv2PairFor", [
       uniswapV2FactoryAddress,
       tokenA,
-      tokenB
-    );
+      tokenB,
+    ]),
+    target: systemContractAddress,
+  }));
+
+  const { returnData } = await multicallContract.callStatic.aggregate(calls);
+
+  const validPairs = returnData
+    .map((data: any, index: number) => {
+      try {
+        const pair = systemContract.interface.decodeFunctionResult(
+          "uniswapv2PairFor",
+          data
+        )[0];
+        return pair !== ethers.constants.AddressZero ? pair : null;
+      } catch {
+        return null;
+      }
+    })
+    .filter((pair: string | null) => pair !== null);
+
+  const pairCalls = validPairs
+    .map((pair: string) => [
+      {
+        callData: new ethers.utils.Interface(
+          UniswapV2Pair.abi
+        ).encodeFunctionData("token0"),
+        target: pair,
+      },
+      {
+        callData: new ethers.utils.Interface(
+          UniswapV2Pair.abi
+        ).encodeFunctionData("token1"),
+        target: pair,
+      },
+      {
+        callData: new ethers.utils.Interface(
+          UniswapV2Pair.abi
+        ).encodeFunctionData("getReserves"),
+        target: pair,
+      },
+    ])
+    .flat();
+
+  const pairReturnData = await multicallContract.callStatic.aggregate(
+    pairCalls
+  );
 
-    if (pair === ethers.constants.AddressZero) return null;
+  const pools = [];
+  const uniswapInterface = new ethers.utils.Interface(UniswapV2Pair.abi);
 
+  for (let i = 0; i < pairReturnData.returnData.length; i += 3) {
+    const pairIndex = Math.floor(i / 3);
+    const pair = validPairs[pairIndex];
+
+    if (
+      !pairReturnData.returnData[i] ||
+      !pairReturnData.returnData[i + 1] ||
+      !pairReturnData.returnData[i + 2]
+    ) {
+      console.warn(`Missing data for pair ${pair} at index ${i}`);
+      continue;
+    }
+
+    const token0Data = pairReturnData.returnData[i];
+    const token1Data = pairReturnData.returnData[i + 1];
+    const reservesData = pairReturnData.returnData[i + 2];
+
+    // Check if data can be decoded
+    let token0, token1, reserves;
+    try {
+      token0 = uniswapInterface.decodeFunctionResult("token0", token0Data)[0];
+    } catch {
+      continue;
+    }
+    try {
+      token1 = uniswapInterface.decodeFunctionResult("token1", token1Data)[0];
+    } catch {
+      continue;
+    }
```

Comment:
We can try/catch in the same block if the granular error is not needed
```suggestion
    try {
      token0 = uniswapInterface.decodeFunctionResult("token0", token0Data)[0];
      token1 = uniswapInterface.decodeFunctionResult("token1", token1Data)[0];
    } catch {
      continue;
    }
```